From 516c31394b186e4f23d5cd688dc7a965e4ab868c Mon Sep 17 00:00:00 2001
From: David Kaspar <dkaspar@redhat.com>
Date: Fri, 10 Jul 2015 10:46:33 +0200
Subject: [PATCH] 'tcsh_posix_status' variable implemented.

  $tcsh_posix_status variable now co-exist with the $anyerror variable
  from the upstream for compatibility reasons.

  NOTE: Added test for $anyerror variable from upstream.
        Added test for $tcsh_posix_status variable.
        Manual page section about 'status' synced with upstream.

  Resolves: #1123854
---
 sh.c               |  2 ++
 sh.h               |  1 +
 sh.proc.c          |  2 +-
 sh.set.c           |  5 +++++
 tc.const.c         |  2 ++
 tcsh.man           | 27 ++++++++++++++++-----------
 tests/variables.at | 50 ++++++++++++++++++++++++++++++++++++++++++++++++++
 7 files changed, 77 insertions(+), 12 deletions(-)

diff --git a/sh.c b/sh.c
index f897317..498bf25 100644
--- a/sh.c
+++ b/sh.c
@@ -354,6 +354,8 @@ main(int argc, char **argv)
     anyerror = 1;		/* for compatibility */
     setcopy(STRanyerror, STRNULL, VAR_READWRITE);
 
+    tcsh_posix_status = 0;
+
     /* Default history size to 100 */
     setcopy(STRhistory, str2short("100"), VAR_READWRITE);
 
diff --git a/sh.h b/sh.h
index 4e3f13c..74b7719 100644
--- a/sh.h
+++ b/sh.h
@@ -597,6 +597,7 @@ EXTERN int    editing IZERO;	/* doing filename expansion and line editing */
 EXTERN int    noediting IZERO;	/* initial $term defaulted to noedit */
 EXTERN int    bslash_quote IZERO;/* PWP: tcsh-style quoting?  (in sh.c) */
 EXTERN int    anyerror IZERO;	/* propagate errors from pipelines/backq */
+EXTERN int    tcsh_posix_status	IZERO;	/* negation for anyerror */
 EXTERN int    compat_expr IZERO;/* csh-style expressions? */
 EXTERN int    isoutatty IZERO;	/* is SHOUT a tty */
 EXTERN int    isdiagatty IZERO;/* is SHDIAG a tty */
diff --git a/sh.proc.c b/sh.proc.c
index 4fd522b..0c5fc25 100644
--- a/sh.proc.c
+++ b/sh.proc.c
@@ -560,7 +560,7 @@ pjwait(struct process *pp)
     do {
 	/* In case of pipelines only the result of the last
 	 * command should be taken in account */
-	if (!anyerror && !(fp->p_flags & PBRACE)
+	if ((!anyerror || tcsh_posix_status) && !(fp->p_flags & PBRACE)
 		&& ((fp->p_flags & PPOU) || (fp->p_flags & PBACKQ)))
 	    continue;
 	if (fp->p_reason)
diff --git a/sh.set.c b/sh.set.c
index 0f98a2b..8d660e0 100644
--- a/sh.set.c
+++ b/sh.set.c
@@ -109,6 +109,9 @@ update_vars(Char *vp)
     else if (eq(vp, STRanyerror)) {
 	anyerror = 1;
     }
+    else if (eq(vp, STRtcsh_posix_status)) {
+	tcsh_posix_status = 1;
+    }
     else if (eq(vp, STRsymlinks)) {
 	Char *pn = varval(vp);
 
@@ -768,6 +771,8 @@ unset(Char **v, struct command *c)
 	loginsh = 0;
     if (adrof(STRanyerror) == 0)
 	anyerror = 0;
+    if (adrof(STRtcsh_posix_status) == 0)
+	tcsh_posix_status = 0;
     if (adrof(STRwordchars) == 0)
 	word_chars = STR_WORD_CHARS;
     if (adrof(STRedit) == 0)
diff --git a/tc.const.c b/tc.const.c
index 641e234..d942542 100644
--- a/tc.const.c
+++ b/tc.const.c
@@ -44,6 +44,8 @@ Char STRrootdefautologout[] = { '1', '5', '\0' };
 Char STRautomatic[]	= { 'a', 'u', 't', 'o', 'm', 'a', 't', 'i', 'c',
 			    '\0' };
 Char STRanyerror[]	= { 'a', 'n', 'y', 'e', 'r', 'r', 'o', 'r', '\0' };
+Char STRtcsh_posix_status[] = {'t', 'c', 's', 'h', '_', 'p', 'o', 's', 'i', 'x',
+			       '_', 's', 't', 'a', 't', 'u', 's', '\0' };
 Char STRhangup[]	= { 'h', 'a', 'n', 'g', 'u', 'p', '\0' };
 Char STRaout[]		= { 'a', '.', 'o', 'u', 't', '\0' };
 Char STRtty[]		= { 't', 't', 'y', '\0' };
diff --git a/tcsh.man b/tcsh.man
index 8cc45c8..f73cb53 100644
--- a/tcsh.man
+++ b/tcsh.man
@@ -4356,17 +4356,22 @@ Reset to 1 in login shells.
 See also \fBloginsh\fR.
 .TP 8
 .B status
-The status returned by the last command, unless the variable
-.B anyerror
-is set, and any error in a pipeline or a backquote expansion will be
-propagated (this is the default
-.B csh
-behavior, and the current
-.B tcsh
-default). If it terminated
-abnormally, then 0200 is added to the status.  Builtin commands
-which fail return exit status `1', all other builtin commands
-return status `0'.
+The exit status from the last command or backquote expansion, or any
+command in a pipeline is propagated to \fBstatus\fR.  (This is also the
+default \fBcsh\fR behavior.)
+This default does not match what POSIX mandates (to return the
+status of the last command only). To match the POSIX behavior, you need
+to unset \fBanyerror\fR.
+.RS +8
+.PP
+If the \fBanyerror\fR variable is unset, the exit status of a pipeline
+is determined only from the last command in the pipeline, and the exit
+status of a backquote expansion is \fInot\fR propagated to \fBstatus\fR.
+.PP
+If a command terminated abnormally, then 0200 is added to the status.
+Builtin commands which fail return exit status `1', all other builtin
+commands return status `0'.
+.RE
 .TP 8
 .B symlinks \fR(+)
 Can be set to several different values to control symbolic link (`symlink')
diff --git a/tests/variables.at b/tests/variables.at
index 424e4da..d2b5c08 100644
--- a/tests/variables.at
+++ b/tests/variables.at
@@ -39,6 +39,28 @@ VAR_UNSET([afsuser])
 VAR_UNSET([ampm])
 
 
+AT_SETUP([$ anyerror])
+
+AT_DATA([exit_status.csh],
+[[echo $?anyerror $anyerror
+false | true ; echo $?
+unset anyerror ; echo $?anyerror
+false | true ; echo $?
+set anyerror
+false | true ; echo $?
+]])
+
+AT_CHECK([tcsh -f exit_status.csh],,
+[1
+1
+0
+0
+1
+])
+
+AT_CLEANUP
+
+
 AT_SETUP([$ argv])
 
 AT_CHECK([tcsh -f -c 'echo $argv; echo $2' foo bar baz], ,
@@ -937,6 +959,34 @@ AT_CHECK([tcsh -f -c 'echo $?tcsh'], ,
 AT_CLEANUP
 
 
+AT_SETUP([$ tcsh_posix_status])
+
+AT_DATA([exit_status.csh],
+[[echo $?tcsh_posix_status
+false | true ; echo $?
+set tcsh_posix_status = 1 ; echo $?tcsh_posix_status $tcsh_posix_status
+false | true ; echo $?
+set tcsh_posix_status = 0 ; echo $?tcsh_posix_status $tcsh_posix_status
+# Note it is still set!
+false | true ; echo $?
+unset tcsh_posix_status ; echo $?tcsh_posix_status
+false | true ; echo $?
+]])
+
+AT_CHECK([tcsh -f exit_status.csh],,
+[0
+1
+1 1
+0
+1 0
+0
+0
+1
+])
+
+AT_CLEANUP
+
+
 AT_SETUP([$ term])
 
 AT_DATA([term.csh],
-- 
1.8.3.1

